The webm videos on this page don't play in Safari (work in Firefox and Chrome)

jyavenard opened the pull request:
#### 7738281a461ecea0aff6612dad9e807ec7deceea
<pre>
The webm videos on this page don&apos;t play in Safari (work in Firefox and Chrome)
<a href="https://bugs.webkit.org/show_bug.cgi?id=304109">https://bugs.webkit.org/show_bug.cgi?id=304109</a>
<a href="https://rdar.apple.com/problem/166439682">rdar://problem/166439682</a>

Reviewed by Youenn Fablet.

Create createVideoInfoFromAV1Stream utility and make computeAV1InputFormat
use it.
Add supports for AV1 in a WebM.

We only enable support for those videos if an AV1 hardware decoder is available.
To be able to determine if HW support is available in the web content process
like in 300722@main, we now distinguish if AV1 HW decoder is available
in process vs globally by adding a av1HardwareDecoderAvailableInProcess method.

Test Added (But disabled as it requires specific hardware on iOS and macOS).

* LayoutTests/media/content/test-av1-709-fullrange.webm: Added.
* LayoutTests/media/media-webm-av1-expected.txt: Added.
* LayoutTests/media/media-webm-av1.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/AV1Utilities.cpp:
(WTF::isValidEnum&lt;WebCore::AV1ConfigurationLevel&gt;): 31 value is valid and seen in a few webm found on the web.
(WebCore::createPlatformVideoColorSpaceFromAV1CodecConfigurationRecord):
(WebCore::createVideoInfoFromAV1CodecConfigurationRecord):
(WebCore::readULEBSize):
(WebCore::getSequenceHeaderOBU):
(WebCore::createVideoInfoFromAV1Stream):
* Source/WebCore/platform/graphics/AV1Utilities.h:
* Source/WebCore/platform/graphics/CodecUtilities.cpp:
(WebCore::humanReadableStringFromCodecString): Extend with 31 (Maximum) value.
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.mm:
(WebCore::av1HardwareDecoderAvailable):
(WebCore::av1HardwareDecoderAvailableInProcess):
(WebCore::readULEBSize): Deleted.
(WebCore::getSequenceHeaderOBU): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnTrackEntry):
(WebCore::WebMParser::OnFrame):
(WebCore::WebMParser::VideoTrackData::consumeFrameData):
(WebCore::WebMParser::isSupportedVideoCodec):
(WebCore::SourceBufferParserWebM::isContentTypeSupported):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
(computeAV1InputFormat): Use existing AV1Utilities/CMUtilities.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_sharedPreferencesForWebProcess):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didInitialize):

Canonical link: <a href="https://commits.webkit.org/304510@main">https://commits.webkit.org/304510@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fac0cdf82f559aabc4b95a49fd2cf71e52d5245

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135823 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8200 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47115 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143536 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4d76c4ad-e9e0-443c-9d20-42ed5a882664) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137692 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8858 "Hash 0fac0cdf for PR 55373 does not build (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8045 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103802 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/22fd3a7d-74f1-44c5-b84f-3ff10be31d68) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138769 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/8858 "Hash 0fac0cdf for PR 55373 does not build (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121741 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84678 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f05ab88e-0455-4eb3-8fca-608fcd784dd5) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/8858 "Hash 0fac0cdf for PR 55373 does not build (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3747 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4142 "Built successfully") | | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/8858 "Hash 0fac0cdf for PR 55373 does not build (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39930 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146286 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7885 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40496 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112158 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112544 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28548 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6014 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118041 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61798 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7932 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7667 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71482 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7889 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7752 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/AV1Utilities.cpp:
```diff
@@ -596,16 +614,758 @@ std::optional<AV1CodecConfigurationRecord> parseAV1DecoderConfigurationRecord(co
+    BitReader bitReader(data);
+    std::optional<size_t> value;
+
+    // seq_profile f(3)
```
What's the strategy for testing and/or fuzzing these binary parsers?

jyavenard replied:
> What's the strategy for testing and/or fuzzing these binary parsers?

all media parsers have been moved to the web content process now.

There's several AV1 tests exercising this code in WPT's VideoDecoder/VideoEncoder.

weinig replied:
I mean, I think its great we have parsers in the web content process, but I don't think that means we don't need testing and fuzzing.

Binary parsers are notoriously hard to get right, so it seems like any that we have, and certainly any new ones, should have good test coverage and some fuzzing.

youennf commented on Source/WebCore/platform/graphics/AV1Utilities.cpp:
```diff
@@ -1150,4 +1152,220 @@ std::optional<AV1CodecConfigurationRecord> parseSequenceHeaderOBU(std::span<cons
+    return colorSpace;
+}
+
+static RefPtr<VideoInfo> createVideoInfoFromAV1CodecConfigurationRecord(const AV1CodecConfigurationRecord& record, std::span<const uint8_t> fullOBUHeader)
```
```suggestion
static Ref<VideoInfo> createVideoInfoFromAV1CodecConfigurationRecord(const AV1CodecConfigurationRecord& record, std::span<const uint8_t> fullOBUHeader)
```

youennf commented on Source/WebCore/platform/graphics/AV1Utilities.cpp:
```diff
@@ -1150,4 +1152,220 @@ std::optional<AV1CodecConfigurationRecord> parseSequenceHeaderOBU(std::span<cons
+{
+    RefPtr videoInfo = VideoInfo::create();
+
+    videoInfo->codecName = { "av01" };
```
We should really stop having a refcounted struct like this.
From what I can see, we should make it a proper class.

jyavenard replied:
It mimics CMFormatDescription

youennf commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1492,9 +1511,13 @@ MediaPlayerEnums::SupportsType SourceBufferParserWebM::isContentTypeSupported(co
         if (supportsLimitedMatroska && (codec.startsWith("avc1."_s) || codec == "pcm"_s))
             continue;
 
+
```
```suggestion
```

jyavenard committed:
The webm videos on this page don't play in Safari (work in Firefox and Chrome)
https://bugs.webkit.org/show_bug.cgi?id=304109
rdar://problem/166439682

Reviewed by Youenn Fablet.

Create createVideoInfoFromAV1Stream utility and make computeAV1InputFormat
use it.
Add supports for AV1 in a WebM.

We only enable support for those videos if an AV1 hardware decoder is available.
To be able to determine if HW support is available in the web content process
like in 300722@main, we now distinguish if AV1 HW decoder is available
in process vs globally by adding a av1HardwareDecoderAvailableInProcess method.

Test Added (But disabled as it requires specific hardware on iOS and macOS).

* LayoutTests/media/content/test-av1-709-fullrange.webm: Added.
* LayoutTests/media/media-webm-av1-expected.txt: Added.
* LayoutTests/media/media-webm-av1.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/AV1Utilities.cpp:
(WTF::isValidEnum<WebCore::AV1ConfigurationLevel>): 31 value is valid and seen in a few webm found on the web.
(WebCore::createPlatformVideoColorSpaceFromAV1CodecConfigurationRecord):
(WebCore::createVideoInfoFromAV1CodecConfigurationRecord):
(WebCore::readULEBSize):
(WebCore::getSequenceHeaderOBU):
(WebCore::createVideoInfoFromAV1Stream):
* Source/WebCore/platform/graphics/AV1Utilities.h:
* Source/WebCore/platform/graphics/CodecUtilities.cpp:
(WebCore::humanReadableStringFromCodecString): Extend with 31 (Maximum) value.
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.mm:
(WebCore::av1HardwareDecoderAvailable):
(WebCore::av1HardwareDecoderAvailableInProcess):
(WebCore::readULEBSize): Deleted.
(WebCore::getSequenceHeaderOBU): Deleted.
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnTrackEntry):
(WebCore::WebMParser::OnFrame):
(WebCore::WebMParser::VideoTrackData::consumeFrameData):
(WebCore::WebMParser::isSupportedVideoCodec):
(WebCore::SourceBufferParserWebM::isContentTypeSupported):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
(computeAV1InputFormat): Use existing AV1Utilities/CMUtilities.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_sharedPreferencesForWebProcess):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didInitialize):

Canonical link: https://commits.webkit.org/304510@main
